### PR TITLE
Full persistence of pods

### PR DESCRIPTION
**Pull request by jasongilanfarr**
TODOs:

- [ ] Add new test cases.

---
This change is [Reviewable](https://reviewable.io/reviews/mesosphere/marathon/4271)

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/json/GroupUpdate.scala [hunk 1]
```diff
@@ -40,7 +41,7 @@ case class GroupUpdate(
         .map(app => app.id -> app)(collection.breakOut)
 
     val effectiveDependencies = dependencies.fold(current.dependencies)(_.map(_.canonicalPath(current.id)))
-    Group(current.id, effectiveApps, effectiveGroups, effectiveDependencies, timestamp)
+    Group(current.id, effectiveApps, current.pods, effectiveGroups, effectiveDependencies, timestamp)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/json/GroupUpdate.scala]: maybe a comment along the lines of "pods aren't considered in dependency calculations"

File: src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala [hunk 2]
```diff
@@ -133,7 +133,8 @@ private[impl] class GroupManagerActor(
         _ = validateOrThrow(plan)(DeploymentPlan.deploymentPlanValidator(config))
         _ = log.info(s"Computed new deployment plan:\n$plan")
         _ <- scheduler.deploy(plan, force)
-        _ <- groupRepo.storeRoot(plan.target, plan.createdOrUpdatedApps, plan.deletedApps)
+        // TODO(PODS): Add in createdOrUpdatedPods and deletedPods
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala]: should we track this with a separate "pods/" GH issue, or will this be addressed in this PR?

File: src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala [hunk 3]
```diff
@@ -173,18 +187,28 @@ private[storage] trait ScanBehavior[K, C, S] { this: FSM[State, Data] with Actor
   def computeActualDeletions(
     appsStored: Set[PathId],
     appVersionsStored: Map[PathId, Set[OffsetDateTime]], // scalastyle:off
+    podsStored: Set[PathId],
+    podVersionsStored: Map[PathId, Set[OffsetDateTime]],
     rootsStored: Set[OffsetDateTime],
-    scanDone: ScanDone): (Set[PathId], Map[PathId, Set[OffsetDateTime]], Set[OffsetDateTime]) = {
-    val ScanDone(appsToDelete, appVersionsToDelete, rootVersionsToDelete) = scanDone
+    scanDone: ScanDone): BlockedEntities = {
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala]: :+1:

File: src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala [hunk 4]
```diff
@@ -269,25 +308,45 @@ private[storage] trait ScanBehavior[K, C, S] { this: FSM[State, Data] with Actor
...
+        usedPods.map { id =>
+          podRepository.versions(id).runWith(Sink.sortedSet).map(id -> _)
+        }
+      }.map(_.filter(_._2.size > maxVersions).toMap)
+    }
+
     async {
       val inUseRootFuture = rootsInUse()
       val allAppIdsFuture = appRepository.ids().runWith(Sink.set)
+      val allPodIdsFuture = podRepository.ids().runWith(Sink.set)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala]: seems like there's a pattern here. useful to refactor into a more abstract form? or maybe later if we add support for other types of persisted things (like v3 assemblies?)

**Comment by jasongilanfarr** [comment]
Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/main/scala/mesosphere/marathon/api/v2/json/GroupUpdate.scala, line 44 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4271#-KQQvtJGE2XiA8zWmQRP-r1-44:-KQQxIiisJE8GPR-n4BI:biyrw3i) ([raw file](https://github.com/mesosphere/marathon/blob/49f5a10dd930ee112e1652ee2faa17995d5ad101/src/main/scala/mesosphere/marathon/api/v2/json/GroupUpdate.scala#L44)):*
<details><summary><i>Previously, jdef (James DeFelice) wrote…</i></summary>
> maybe a comment along the lines of "pods aren't considered in dependency calculations"
</details>
Haven't gone and touched the dependencies yet. That's part of the group API changes.

---

*[src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala, line 136 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4271#-KQQvtJGE2XiA8zWmQRQ-r1-136:-KQQxOOuepbynfPDElDU:bz287u5) ([raw file](https://github.com/mesosphere/marathon/blob/49f5a10dd930ee112e1652ee2faa17995d5ad101/src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala#L136)):*
<details><summary><i>Previously, jdef (James DeFelice) wrote…</i></summary>
> should we track this with a separate "pods/" GH issue, or will this be addressed in this PR?
</details>
This should be covered by DeploymentPlan integration. I don't think it should be in this change.

---

*[src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala, line 322 \[r1\]](https://reviewable.io:443/reviews/mesosphere/marathon/4271#-KQQvtJHbwDFkD-FC6eg-r1-322:-KQQxUbb4uGzxEfaUDlX:b34jem9) ([raw file](https://github.com/mesosphere/marathon/blob/49f5a10dd930ee112e1652ee2faa17995d5ad101/src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala#L322)):*
<details><summary><i>Previously, jdef (James DeFelice) wrote…</i></summary>
> seems like there's a pattern here. useful to refactor into a more abstract form? or maybe later if we add support for other types of persisted things (like v3 assemblies?)
</details>
There's definitely a pattern, but I don't think its worth refactoring _yet_

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mesosphere/marathon/4271)*

**Comment by jdef** [comment]
:lgtm:

---

Review status: 0 of 28 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mesosphere/marathon/4271#-:-KQRt4VceWo6A5u4o20y:bnfp4nl)*

**Pull request merged**